### ISSUE TITLE
setup.py on Python 3 does not install data files that the tests need

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7876]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7876|
|Type|defect|
|Created|2015-04-29 06:55:54Z|
|Branch|https://github.com/twisted/twisted/tree/externaltestfiles-py3-7876|

setup.py install does not install all of the correct data files into a Python 3 install, despite them being included in the tarball. This should be fixed.

This makes the tests pass in a Py3 venv, of which I will make a builder for soon.

<details><summary>Searchable metadata</summary>

```
trac-id__7876 7876
type__defect defect
reporter__hawkowl hawkowl
priority__normal normal
milestone__ 
branch__branches_externaltestfiles_py3_7876 branches/externaltestfiles-py3-7876
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1430290554175377 1430290554175377
changetime__1432187207221708 1432187207221708
version__None None
owner__hawkowl hawkowl
cc__radix
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

(In [44619]) Branching to sdist-7876.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Why?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

And what is "The Right Thing"?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

It turns out that some things that I thought it was not doing were actually the python 3 setup.py not installing files (rather than sdist not adding them) so rather than the ticket this would unblock, I'll make it the ticket that it should be.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

(In [44852]) Branching to externaltestfiles-py3-7876.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl commented|
|-|-|

Okay, so -- sdist generates a correct tarball, this is good. However, the missing piece isn't quite sdist, but that setup.py on Python 3 doesn't actually install all the files.

So this is now *not* sdist, but setup.py. (I am seemingly not very good at debugging).

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl commented|
|-|-|

setup.py in this branch installs the required data files on Python 3. There's also direct tests for the parts that process the file list for distutils.

Builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_9"></a>|@adiroiban set owner to @hawkowl|
|-|-|

Changes look good.

Maybe we can start with adding the virtual env buildslave so that we can test these changes on a slave

I will be happy to review/help the changes required to create the new py3 venv builder

----------

I am not happy with keeping another manual list of files for python3, but I think that this can be done in a separate ticket.

I think that the best practice is to not have these files at all, but rather include them in the test_.py module and write them into temporary files at test runtime.

Maybe we can put all test data files in the same folder and update setup.py to automatically copy at install time all files from that folder.

What do you say?

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_10"></a>|@hawkowl commented|
|-|-|

I'm going to do a builder for this soon, so that this works and keeps working in the future.

The manual list is unfortunately I think the best solution right now -- eventually this will ~all go away~. Eventually.

Will merge.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44887]) Merge externaltestfiles-py3-7876: setup.py now installs the data files that the tests need on Py3

Author: hawkowl
Reviewer: adiroiban
Fixes: [#7876](https://github.com/twisted/twisted/issues/7876)
